File: Structural/Lab13/C#/Entities/CachedAgency.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Interfaces;

namespace C_.Entities
{
    public class CachedAgency : IAgency
    {
        private Agency _agency;
        private Dictionary<int, RealEstate> _cache = new Dictionary<int, RealEstate>();
        public CachedAgency(Agency agency)
        {
            _agency = agency;
        }

        public RealEstate FindEstateByID(int id){
            if (_cache.ContainsKey(id)){
                Console.WriteLine($"Get from cache {id}");
                return _cache[id];
            }
            else
            {
                Console.WriteLine($"Get from base {id}");
                var realEstate = _agency.FindEstateByID(id);
                _cache.Add(id, realEstate);
                return realEstate;
            }
        }
    }
}

File: Structural/Lab13/Typescript/src/Entities/cached_agency.ts
import { IAgency } from "../Interfaces/agency";
import { Agency } from "./agency";
import { RealEstate } from "./real_estate";

export class CachedAgency implements IAgency {
    
    private _agency: Agency;
    private _cache: { [id: number]: RealEstate } = {};
    constructor(agency: Agency) {
        this._agency = agency;
    }

    public FindEstateByID(id: number): RealEstate {
        if (this._cache.hasOwnProperty(id)) {
            console.log(`Get from cache ${id}`);
            return this._cache[id];
        } else {
            console.log(`Get from base ${id}`);
            const realEstate = this._agency.FindEstateByID(id);
            this._cache[id] = realEstate;
            return realEstate;
        }
    }
}